### PR TITLE
distribute/gather functions for `Vector`

### PR DESCRIPTION
**Pull request by dreamer2368**
This implements `Vector::distribute` and `Vector::gather` analogous to `Matrix` class.

- `Vector::distribute` distributes a non-distributed vector into a locally distributed vector with specified local dimension.
- `Vector::gather` gathers a locally distributed vector into a non-distributed vector.

### CODE REVIEW

File: lib/linalg/Vector.cpp [hunk 1]
```diff
@@ -577,6 +584,64 @@ double Vector::localMin(int nmax)
     return v;
 }
 
+void
+Vector::distribute(const int &local_dim)
```
Comment (dylan-copeland) [inline lib/linalg/Vector.cpp:588]: Pass by value?

File: lib/linalg/Vector.cpp [hunk 2]
```diff
@@ -577,6 +584,64 @@ double Vector::localMin(int nmax)
...
+{
+    CAROM_VERIFY(!distributed());
+    CAROM_VERIFY(d_owns_data);
```
Comment (dylan-copeland) [inline lib/linalg/Vector.cpp:591]: Verify that `local_dim >= 0`?

File: lib/linalg/Vector.h [hunk 3]
```diff
@@ -759,6 +759,27 @@ class Vector
...
+     * @brief Distribute this vector among MPI processes,
+     * based on the specified local dimension.
+     * This becomes distributed after this function is executed.
+     *
+     * @pre !distributed()
+     * @pre d_owns_data
+     *
+     * @param[in] local_dim dimension for local MPI rank.
+     */
+    void distribute(const int &local_dim);
```
Comment (dylan-copeland) [inline lib/linalg/Vector.h:772]: These functions would be more general if an MPI communicator is an argument, with default value `MPI_COMM_WORLD`.

Comment (dylan-copeland) [reply lib/linalg/Vector.h:772]: On second thought, `MPI_COMM_WORLD` is used in other places in the class, so the `Vector` constructor should take an optional argument for the communicator. It's not really necessary to do this in this PR, but it's a possibility.

File: lib/linalg/Vector.h [hunk 4]
```diff
@@ -759,6 +759,27 @@ class Vector
...
+
+    /**
+     * @brief Gather all the distributed elements among MPI processes.
+     * This becomes not distributed after this function is executed.
```
Comment (dylan-copeland) [inline lib/linalg/Vector.h:776]: Further comment that the data is identical on all processes after this operation?

File: lib/linalg/Vector.h [hunk 5]
```diff
@@ -759,6 +759,27 @@ class Vector
...
 
+    /**
+     * @brief Distribute this vector among MPI processes,
+     * based on the specified local dimension.
+     * This becomes distributed after this function is executed.
+     *
+     * @pre !distributed()
+     * @pre d_owns_data
+     *
+     * @param[in] local_dim dimension for local MPI rank.
```
Comment (dylan-copeland) [inline lib/linalg/Vector.h:770]: Comment that `local_dim` must sum to the current dimension over all processes?

File: lib/linalg/Vector.h (hunk 3 above)
Comment (dreamer2368) [reply lib/linalg/Vector.h]: It was also one of my wish, that libROM parallel linear algebra should support custom MPI communicator. I attempted it a while ago, but it wasn't as simple as I thought, particularly in the interface with scalapack. We can discuss about this more.

For now, in order to be consistent with the current practice throughout the code, I kept `MPI_COMM_WORLD`.

**Review by dylan-copeland — APPROVED** [review]

**Review by chldkdtn — APPROVED** [review]

**Review by ckendrick — APPROVED** [review]

**Pull request merged**